src/index/CompressedBlockPrefiltering.h

```diff
@@ -0,0 +1,140 @@
+//  Copyright 2024, University of Freiburg,
+//                  Chair of Algorithms and Data Structures
+//  Author: Hannes Baumann <baumannh@informatik.uni-freiburg.de>
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "global/Id.h"
+#include "global/ValueIdComparators.h"
+#include "index/CompressedRelation.h"
+
+// The compressed block metadata (see `CompressedRelation.h`) that we use to
+// filter out the non-relevant blocks by checking their content of
+// `firstTriple_` and `lastTriple_` (`PermutedTriple`)
+using BlockMetadata = CompressedBlockMetadata;
+
+namespace prefilterExpressions {
+//______________________________________________________________________________
+/*
+`PrefilterExpression` represents a base class for the following sub-classes that
+implement the block-filtering procedure for the specific relational + logical
+operations/expressions.
+
+Remark: We do not actually evaluate the respective SPARQL Expression. We only
+pre-filter w.r.t. blocks that contain relevant data for the actual evaluation of
+those expressions to make the evaluation procedure more efficient.
+
+The block-filtering is applied with the following operations:
+Relational Expressions - `<=`, `>=`, `<`, `>`, `==` and `!=`
+Logical Operations - `and` and `or`
+*/
+
+class PrefilterExpression {
+ public:
+  // The respective metadata to the blocks is expected to be provided in
+  // a sorted order (w.r.t. the relevant column).
+  virtual std::vector<BlockMetadata> evaluate(std::vector<BlockMetadata>& input,
+                                              size_t evaluationColumn) = 0;
```

Comment:
The interface looks odd, either we filter in place (then the return type is `void`, or we return a new vector
(then the argument can be a `const vector&`.
Also this function can probably be const in all cases.